src/auth/linking.ts

```diff
@@ -0,0 +1,242 @@
+import aes from "keystore-idb/lib/aes/index.js"
+import config from "keystore-idb/lib/config.js"
+import rsa from "keystore-idb/lib/rsa/index.js"
+import utils from "keystore-idb/lib/utils.js"
+import { KeyUse, SymmAlg, HashAlg, CharSize } from "keystore-idb/lib/types.js"
+import * as did from "../did/index.js"
+import * as ucan from "../ucan/index.js"
+import { publishOnChannel } from "./index.js"
+import * as auth from "./index.js"
+
+type LinkingRole = "CONSUMER" | "PRODUCER"
+
+type LinkingStep = "BROADCAST" | "NEGOTIATION" | "DELEGATION"
+
+type LinkingState = {
+  sessionKey: CryptoKey | null
+  temporaryRsaPair: CryptoKeyPair | null
+  role: LinkingRole | null
+  step: LinkingStep | null
+}
+
+const ls: LinkingState = {
+  sessionKey: null,
+  temporaryRsaPair: null,
+  role: null,
+  step: null
+}
+
+const resetLinkingState = () => {
+  ls.sessionKey = null
+  ls.temporaryRsaPair = null
+}
+
+const nextStep = () => {
+  switch (ls.step) {
+    case "BROADCAST":
+      ls.step = "NEGOTIATION"
+      break
+    case "NEGOTIATION":
+      ls.step = "DELEGATION"
+      break
+    default:
+      ls.step = "BROADCAST"
+  }
+}
+
+export const startLinkingConsumer = async (username: string): Promise<null> => {
+  ls.role = "CONSUMER"
+  ls.step = "BROADCAST"
+  await auth.openChannel(username)
+  await sendTemporaryExchangeKey()
+  return null
+}
+
+export const startLinkingProducer = async (username: string): Promise<null> => {
+  ls.role = "PRODUCER"
+  ls.step = "BROADCAST"
+  await auth.openChannel(username)
+  return null
+}
```

Comment:
I'm 100% open to suggestions here, I'm not sure I love this. I have these as separate, explicit functions because the get called separately (Producer on whatever linking UI page when authenticated, and Consumer when the user has provided user info and explicitly started the process). I suppose these two should be injectable- although both get called directly from implementer code and it might be easier to just re-implement with any modification (I.e. if you're not using usernames, etc ... and you probably have already swapped in different channel stuff).

TODO: Need to figure out how to handle the challenge PINs (since that all happens over the channel) .. I'm leaning towards having these functions have a callback for `handleChallengePin` (where both sides display the PIN, and producer would have an accept/reject mechanism).